python/lsst/pipe/tasks/assembleCoadd.py

```diff
@@ -309,8 +350,54 @@ def __init__(self, *args, **kwargs):
 
         self.warpType = self.config.warpType
 
+    @classmethod
+    def getOutputDatasetTypes(cls, config):
+        outputTypeDict = super().getOutputDatasetTypes(config)
+        if not config.doNImage:
+            outputTypeDict.pop("nImage", None)
+        return outputTypeDict
+
+    @classmethod
+    def getInputDatasetTypes(cls, config):
+        inputTypeDict = super().getInputDatasetTypes(config)
+        if not config.doMaskBrightObjects:
+            inputTypeDict.pop("brightObjectMask", None)
+        else:  # TODO: remove this else branch after DM-17300
+            raise RuntimeError("Gen3 Bright object masks unavailable. Set config doMaskBrightObjects=False")
+        return inputTypeDict
+
+    def adaptArgsAndRun(self, inputData, inputDataIds, outputDataIds, butler):
```

Comment:
Adding a substantial docstring. `adaptArgsAndRun` is the `runDataRef` of Gen3 world. I was hoping this would be called `runQuantum`, but what's now`runQuantum` turned into a beast that's impossible to override. I'm hoping after January 31st we have a design review and can change some of these names.